src/perform_step/symplectic_perform_step.jl

```diff
@@ -74,10 +75,36 @@ end
   f.f2(ku,du,u,p,t)
 end
 
-function initialize!(integrator,cache::C) where
-    {C<:Union{VelocityVerletCache,Symplectic2Cache,Symplectic3Cache,Symplectic4Cache,
-              Symplectic45Cache,Symplectic5Cache,Symplectic6Cache,Symplectic62Cache,
-              McAte8Cache,KahanLi8Cache,SofSpa10Cache}}
+const CachesInpHamilton = Union{Symplectic2Cache,Symplectic3Cache,
+Symplectic4Cache,Symplectic45Cache,Symplectic5Cache,
+Symplectic6Cache,Symplectic62Cache,
+McAte8Cache,KahanLi8Cache,SofSpa10Cache,}
+const CachesInpNewton = Union{VelocityVerletCache,}
+
+const CachesNipHamilton = Union{Symplectic2ConstantCache,Symplectic3ConstantCache,
+Symplectic4ConstantCache,Symplectic45ConstantCache,Symplectic5ConstantCache,
+Symplectic6ConstantCache,Symplectic62ConstantCache,
+McAte8ConstantCache,KahanLi8ConstantCache,SofSpa10ConstantCache,}
+const CachesNipNewton = Union{VelocityVerletConstantCache,}
+
+# some of the algorithms are designed only for the case
+# f.f2(p, q, pa, t) = p which is the Newton/Lagrange equations
+# If called with different functions (which are possible in the Hamiltonian case)
+# an exception is thrown to avoid silently calculate wrong results.
+verify_f2(::C, f, p, q, pa, t) where C<:CachesNipHamilton = f(p, q, pa, t)
+verify_f2(::C, f, res, p, q, pa, t) where C<:CachesInpHamilton = f(res, p, q, pa, t)
+
+function verify_f2(::C, f, p, q, pa, t) where C<:CachesNipNewton
+    res = f(p, q, pa, t)
+    res == p ? p : throwex()
+end
+function verify_f2(::C, f, res, p, q, pa, t) where C<:CachesInpNewton
+    f(res, p, q, pa, t)
+    res == p ? res : throwex()
+end
+throwex() = throw(ArgumentError("This algorithm is invalid if f2(p, q, t) != p"))
```

Comment:
which methods throw this argument? That would be good to document. It looks like most are fixed?

Reply:
There is only one algorithm `VelocityVerletCache`, which I could not generalize to the Hermitian case, because it uses the time derivative of velocity.

Reply:
I could put the name of the algorithm found in `integrator.alg` into the message of the exception.

Reply:
What I meant was that I think that it should get a mention on https://diffeq.sciml.ai/stable/solvers/dynamical_solve/#Symplectic-Integrators